Sources/JWTKit/RSA/PrimeGenerator.swift

```diff
@@ -0,0 +1,129 @@
+import Foundation
+
+struct PrimeGenerator {
+    /// The following algorithm recovers the prime factors of a modulus, given the public and private exponents. 
+    /// The algorithm is based on Fact 1 in [Boneh 1999].
+    public static func calculatePrimeFactors(n: BigInt, e: BigInt, d: BigInt) throws -> (p: BigInt, q: BigInt) {
+        let k = (d * e) - 1
+
+        guard k % 2 == 0 else {
+            throw RSAError.keyInitializationFailure
+        }
+
+        var r = k
+        var t = 0
+
+        repeat {
+            r = r / 2
+            t += 1
+        } while r % 2 == 0
+
+        var y: BigInt = 0
+        var i = 1
+
+        // If the prime factors are not revealed after 100 iterations, 
+        // then the probability is overwhelming that the modulus is not the product of two prime factors, 
+        // or that the public and private exponents are not consistent with each other.
+        while i <= 100 {
+            let g = BigInt.randomInteger(lessThan: n - 1)
+            y = g.power(r, modulus: n)
+
+            guard y != 1 && y != n - 1 else {
+                continue
+            }
+
+            var j = 1
+            var x: BigInt
+
+            while j <= t - 1 {
+                x  = y.power(2, modulus: n)
+
+                guard x != 1 else {
+                    break
+                }
+
+                guard x != n - 1 else {
+                    continue
+                }
+
+                y = x
+                j += 1
+            }
+
+            x = y.power(2, modulus: n)
+            if x == 1 {
+                let p = (y - 1).gcd(with: n)
+                let q = n / p
+
+                return (p, q)
+            }
+            i += 1
+        }
+
+        throw RSAError.keyInitializationFailure
+    }
+}
+
+extension BigInt {
+    public static func randomInteger(lessThan n: BigInt) -> BigInt {
+        let bitLength = n.bitWidth - n.leadingZeroBitCount
+        var random: BigInt
+        repeat {
+            random = BigInt.randomInteger(withExactWidth: bitLength)
+        } while random >= n
+        return random
+    }
+
+    public static func randomInteger(withExactWidth width: Int) -> BigInt {
+        let byteCount = (width + 7) / 8
+        var random = BigInt()
+        
+        let bytes = (0..<byteCount).map { _ in UInt8.random(in: 0...UInt8.max) }
+        for (index, byte) in bytes.enumerated() {
+            let shiftAmount = 8 * (byteCount - index - 1)
+            random |= BigInt(byte) << shiftAmount
+        }
+        
+        return random
+    }
+    
+    var leadingZeroBitCount: Int {
+        var count = 0
+        var number = self
+        while number > 0 {
+            number /= 2
+            count += 1
+        }
+        return Swift.max(0, bitWidth - count)
```

Comment:
```suggestion
        var n = bitWidth
        for w in self.words.reversed() { // N.B.: reversed() is O(1) due to `ReversedCollection`
            n &-= w.leadingZeroBitCount
            if w > 0 { break }
        }
        return Swift.max(0, n)
```
(Looping over `&-=` N/8 times is a _HELL_ of a lot faster than looping over `/= 2` (or even `&>>= 1`, which would've been massively better but still not as good) N times, not to mention `n.leadingZeroBitCount` will compile down to `__builtin_clz()`.)